### ISSUE TITLE
The current snippet is left when opening an unfocused floating window

### ISSUE THREAD

**Issue by camilledejoye** [SirVer/ultisnips#1391]
**~Expected~ Desired behavior:**
Keep the snippet "alive" until we "really" left the buffer.

**Actual behavior:**
When a floating window opens while completing a snippet it's cleared.

**Steps to reproduce**
The setup might get a bit long, since I first want to have a feedback on the issue I have, I'll abstain to provide one for now.

1. Trigger a snippet with at least two placeholders
2. At anytime during the completion of the snippet, open a floating window. Examples I encoutered:
  2.1. Using [nvim-cmp](https://github.com/hrsh7th/nvim-cmp): select a choice in the completion list while the documentation is enabled (use a floating window to show some detail about the completion item)
  2.2. Call `lua vim.lsp.buf.signature_helper()` from an insert mode mapping
3. Try to jump to the next place holder, it will fail

**Explanaion**
When a floating window is open the `BufEnter` event is triggered for the window's buffer.
This trigger the auto command `vim_helper.command("autocmd BufEnter * call UltiSnips#LeavingBuffer()")` which "exit" the current snippet.
I assume this auto command is used to ensure that no matter how we leave the current buffer we clean up behind ourselves.

This logic does not work well with floating window because nowadays they are used by many plugins to show contextual information during the edition of the current buffer.

I guess a solution would be to redefine the need of "cleaning up".
Maybe it would be acceptable to be more conservative and only clean a snippet when we "leave it from the same buffer".
It would even allow to:
* trigger a snippet
* leave insert/select mode
* switch to another window/tab/buffer to check something, copy something, etc.
* go back to the originally edited buffer, go back in insert mode and continue from where we stopped

Unfortunately, I don't know UltiSnips nor the reason behind the current implementation to be know if it's a realistic option or not.
And I assume it would require a non negligible amount of work to switch to this idea of remembering the snippets by buffer instead of "globally".

I played with some other ideas with more or less success, the one which seems to be the easiest to implement and less prone to weird behavior is to make sure that we are not in insert mode when entering the buffer.
Experimenting with `vim_helper.command("autocmd BufEnter * if 'i' != mode() | call UltiSnips#LeavingBuffer() | endif")` has shown good results until know.
It feels more like a hack than an actual solution though, the main idea being "If I'm still editing then I could not have leave the snippet".

Bottom line I would be curious about your thoughts on the subject, maybe someone has already found some kind of workaround.
Personally I can't use UltiSnips anymore if I want to keep my "contextual aid", after using for more than two years now I will be sad if we couldn't figure out a solution.

-----

- **Operating System**: Arch Linux 5.14.3
- **Vim Version**: NVIM v0.6.0-dev+254-g413e86869
Build type: RelWithDebInfo
- **UltiSnips Version**: 53e1921e3ef015ef658e540c0aa9c4835f9c18a6 (current master)
- **Python inside Vim**: None / 3.9.7
- **Docker repo/vimrc**: None

**Comment by SirVer** [comment]
Fixed for the reported scenario by PR #1417 (commit 6847c3d, merged 2022-01-24). `UltiSnips#LeavingBuffer()` now skips the snippet teardown when the current window is a neovim floating window, so opening an unfocused docs popup with `enter=v:false` (nvim-cmp completion docs, `vim.lsp.buf.signature_help`, copilot suggestions, …) no longer clears the active snippet.

Closing.
